tools/crm_mon.c

```diff
@@ -2807,18 +2853,363 @@ print_failed_actions(FILE *stream, pe_working_set_t *data_set)
+
+    if ((run_at_s) && (run_at_s[0] != 0)) {
+        run_at_s[strlen(run_at_s)-1] = 0; /* Overwrite the newline */
+    }
+
+    switch(output_format) {
+        case mon_output_xml:
+            fprintf(stream, "        <fence_event target=\"%s\" action=\"%s\"",
+                    event->target, event->action);
+            switch(event->state) {
+                case st_done:
+                    fprintf(stream, " state=\"success\"");
+                    break;
+                case st_failed:
+                    fprintf(stream, " state=\"failed\"");
+                    break;
+                default:
+                    fprintf(stream, " state=\"pending\"");
+            }
+            fprintf(stream, " origin=\"%s\" client=\"%s\"",
+                    event->origin, event->client);
+            if (event->delegate) {
+                fprintf(stream, " delegate=\"%s\"", event->delegate);
+            }
+            switch(event->state) {
+                case st_done:
+                case st_failed:
+                    fprintf(stream, " completed=\"%s\"", run_at_s?run_at_s:"");
+                    break;
+                default:
+                    fprintf(stream, " state=\"pending\"");
+            }
+            fprintf(stream, " />\n");
+            break;
+
         case mon_output_plain:
         case mon_output_console:
-            print_as("\n");
+            switch(event->state) {
+                case st_done:
+                    print_as("* %s of %s successful: delegate=%s, client=%s, origin=%s,\n"
+                             "    %s='%s'\n",
+                             action_s, event->target,
+                             event->delegate ? event->delegate : "",
+                             event->client, event->origin,
+                             ((!fence_full_history) && (output_format != mon_output_xml))?
+                             "completion-last-successful":"completed",
+                             run_at_s?run_at_s:"");
+                    break;
+                case st_failed:
+                    print_as("* %s of %s failed: delegate=%s, client=%s, origin=%s,\n"
+                             "    %s='%s'\n",
+                             action_s, event->target,
+                             event->delegate ? event->delegate : "",
+                             event->client, event->origin,
+                             ((!fence_full_history) && (output_format != mon_output_xml))?
+                             "completion-last-failed":"completed",
+                             run_at_s?run_at_s:"");
+                    break;
+                default:
+                    print_as("* %s of %s pending: client=%s, origin=%s\n",
+                             action_s, event->target,
+                             event->client, event->origin);
+            }
+            break;
+
+        default:
+            /* no support for fence history for other formats so far */
```

Comment:
Shouldn't be difficult to add an HTML list

Reply:
Nope - just wanted to be able to get feedback on the xml/console/plain-stuff first.